lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -111,31 +125,44 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module) {
   std::sort(deviceMemory.begin(), deviceMemory.end(), sortMostMemory);
 
   // Try to add functions to devices in order from largest to smallest.
+  std::map<BackendKind, size_t> startPos;
   for (unsigned i = 0; i < logicalDeviceSize.size(); i++) {
```

Comment:
I'm not sure I understand the logic here. But I remember that in the preliminary version of this PR that I tested there was a bug in this loop related to how deviceIds were looked up. So, please re-check the logic just in-case.

Reply:
It is different from the version I sent to you before. The basic idea is, we still sort the logical devices and physical devices based on the available memory size from high to low. And then for each logical device, we select the physical device who has the same backendKind as the logical one and the largest available memory.